Stabilize support for Jupyter Notebooks (#12878)
astral-sh/ruff: docs/configuration.md

### Default inclusions

By default, Ruff will discover files matching `*.py`, `*.pyi`, `*.ipynb`, or `pyproject.toml`.

To lint or format files with additional file extensions, use the [`extend-include`](settings.md#extend-include) setting.

> You can also change the default selection using the [`include`](settings.md#include) setting.

Author: Any suggestion on what can be use for this example? I'm worried that it _might_ signal to the users that we support that file extension. Using `*.ipynb` was ok because we do support it. This is the reason I've removed this example.

Reviewer: `extend-include` has an example with `.pyw` files. Although I think the example isn't needed because we already link to `extend-include` which has an example.

Author: Yeah, I think not having an example here is fine.

## Jupyter Notebook discovery

Ruff has built-in support for linting and formatting [Jupyter Notebooks](https://jupyter.org/).

!!! info
    Notebooks are linted and formatted by default from Ruff version `0.6.0` onwards which marks
    Jupyter Notebook support as stable. You can opt-out of linting and formatting notebooks by

>     adding `*.ipynb` to [`extend-exclude`](settings.md#extend-exclude).

Reviewer: I wonder if we should change the note to: For versions older than `0.6.0` linting and formatting requires explicit opt-in by adding `*.ipynb` to [`extend-include`]

And we may want to add an example of how to exclude notebooks altogether.

Reviewer: Looking at the ecosystem results, an example for how to disable some rules for notebooks only might help adoption.

Author: These are good suggestions. Let me make some edits.

## Jupyter Notebook discovery

Ruff has built-in support for linting and formatting [Jupyter Notebooks](https://jupyter.org/).

!!! info

>     Jupyter Notebook support is marked as stable from Ruff version `0.6.0` onwards and will
>     be linted and formatted by default.

Reviewer: The back reference of the second subsentence is unclear. It should refer to Jupyter files but the first sentence only mentions notebook support
```suggestion
    Jupyter Notebook support is marked as stable from Ruff version `0.6.0` onwards and Jupyter Notebooks are linted and formatted by default.
```

If you'd prefer to either only lint or only format Jupyter Notebook files, you can use the
section specific `exclude` option to do so. For example, the following would only lint Jupyter
Notebook files and not format them:

=== "pyproject.toml"

    ```toml
    [tool.ruff.format]
    exclude = ["*.ipynb"]
    ```

=== "ruff.toml"

    ```toml
    [format]
    exclude = ["*.ipynb"]
    ```

And, conversely, the following would only format Jupyter Notebook files and not lint them:

=== "pyproject.toml"

    ```toml
    [tool.ruff.lint]
    exclude = ["*.ipynb"]
    ```

=== "ruff.toml"

    ```toml
    [lint]
    exclude = ["*.ipynb"]
    ```

You can completely disable Jupyter Notebook support by updating the

> [`extend-exclude`](settings.md#extend-exclude) settings:

Reviewer: ```suggestion
[`extend-exclude`](settings.md#extend-exclude) setting:
```

=== "pyproject.toml"

    ```toml
    [tool.ruff]
    extend-exclude = ["*.ipynb"]
    ```

=== "ruff.toml"

    ```toml
    extend-exclude = ["*.ipynb"]
    ```

If you'd like to ignore certain rules specifically for Jupyter Notebook files, you can do so by
using the [`per-file-ignores`](settings.md#per-file-ignores) setting:

=== "pyproject.toml"

    ```toml
    [tool.ruff.lint.per-file-ignores]
    "*.ipynb" = ["T20"]
    ```

=== "ruff.toml"

    ```toml
    [lint.per-file-ignores]
    "*.ipynb" = ["T20"]
    ```

> There are certain rules that has different behavior when applied to Jupyter Notebook files. For

Reviewer: ```suggestion
Some rules that have different behavior when applied to Jupyter Notebook files. For
```

=== "pyproject.toml"

    ```toml
    [tool.ruff.lint]
    exclude = ["*.ipynb"]
    ```

=== "ruff.toml"

    ```toml
    [lint]
    exclude = ["*.ipynb"]
    ```

Alternatively, passing the notebook file(s) to `ruff` on the command-line will always lint or format
the notebook. For example, `ruff check /path/to/notebook.ipynb` will always lint `notebook.ipynb`.

> Similarly, `ruff format /path/to/notebook.ipynb` will always format `notebook.ipynb`.

Reviewer 2: I think we won't need this section anymore. we can move this section to another place that applies for all file types not just .ipynb
